File: Medium/1414-Find the Minimum Number of Fibonacci Numbers Whose Sum Is K/Greedy.rb
# @param {Integer} k
# @return {Integer}
def find_min_fibonacci_numbers(k)
  nums = [1, 1]
  ret = 0

  nums.push(nums[-2] + nums[-1]) while nums[-1] < k

  while k > 0
    nums.pop while nums[-1] > k
    k -= nums[-1]
    ret += 1
  end

  ret
end

File: Medium/1414-Find the Minimum Number of Fibonacci Numbers Whose Sum Is K/Greedy.rs
impl Solution {
    pub fn find_min_fibonacci_numbers(mut k: i32) -> i32 {
        let mut nums = vec![1, 1];
        let mut i = 1;
        let mut ret = 0;

        while nums[i] < k {
            nums.push(nums[i - 1] + nums[i]);
            i += 1;
        }

        while k > 0 {
            while nums[i] > k {
                i -= 1;
            }
            k -= nums[i];
            ret += 1;
        }

        ret
    }
}
